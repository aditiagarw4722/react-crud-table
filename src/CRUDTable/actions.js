// @flow

export const changeSort = (field: string, direction: string): Object => (
  state: Object,
  props: Object
): Object => {
  const { sort } = state;
  return {
    sort: {
      field,
      direction,
    },
  };
};

export const changePage = (activePage: number): Object => (
  state: Object,
  props: Object
): Object => ({
  pagination: {
    ...state.pagination,
    activePage,
  },
});

export const addRule = (rule: Object): Object => (
  state: Object,
  props: Object
): Object => ({
  queryRules: [...state.queryRules, rule],
});

export const removeRule = (rule: Object): Object => (
  state: Object,
  props: Object
): Object => ({
  queryRules: state.queryRules.filter(
    (x) => x.field !== rule.field || x.condition !== rule.condition
  ),
});
