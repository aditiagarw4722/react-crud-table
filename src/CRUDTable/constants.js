export const SORT_DIRECTIONS = {
  DESCENDING: 'descending',
  ASCENDING: 'ascending',
};

export const ID_FIELD = 'id';

export const FIELDS_COMPONENT_TYPE = 'CRUDTable_Fields';
export const FIELD_COMPONENT_TYPE = 'CRUDTable_Field';
export const CREATE_FORM_COMPONENT_TYPE = 'CRUDTable_CreateForm';
export const DELETE_FORM_COMPONENT_TYPE = 'CRUDTable_DeleteForm';
export const UPDATE_FORM_COMPONENT_TYPE = 'CRUDTable_UpdateForm';
export const PAGINATION_COMPONENT_TYPE = 'CRUDTable_Pagination';

export default {
  SORT_DIRECTIONS,
  ID_FIELD,
  FIELDS_COMPONENT_TYPE,
  FIELD_COMPONENT_TYPE,
  CREATE_FORM_COMPONENT_TYPE,
  DELETE_FORM_COMPONENT_TYPE,
  UPDATE_FORM_COMPONENT_TYPE,
  PAGINATION_COMPONENT_TYPE,
};
