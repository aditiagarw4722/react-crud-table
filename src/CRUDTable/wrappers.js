import bcc from 'bem-react-component-creator';

const { block, element } = bcc('crud-table');

export const Table = block('table');

Table.displayName = 'Table';

Table.Row = element('tr', 'row');
Table.Row.displayName = 'Row';

Table.HeaderCell = element('th', 'header-cell');
Table.HeaderCell.displayName = 'HeaderCell';

Table.Header = element('thead', 'header');
Table.Header.displayName = 'Header';

Table.Body = element('tbody', 'body');
Table.Body.displayName = 'Body';

Table.Cell = element('td', 'cell');
Table.Cell.displayName = 'Cell';

Table.CellLabel = element('div', 'cell-label');
Table.CellLabel.displayName = 'CellLabel';

Table.Caption = element('div', 'caption');
Table.Caption.displayName = 'Caption';

export default { Table };
