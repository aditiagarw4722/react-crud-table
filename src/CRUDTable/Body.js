import React from 'react';
import PropTypes from 'prop-types';

import { getTableFieldValue } from './helpers';
import { Table } from './wrappers';
import Button from '../Button';

const Body = ({
  fields,
  items,
  onDeleteClick,
  onUpdateClick,
  actionsLabel,
  updateTrigger,
  deleteTrigger,
}) => (
  <Table.Body>
    {items.map((item) => (
      <Table.Row key={item.id}>
        {fields.map((field) => (
          <Table.Cell key={field.name}>
            <Table.CellLabel>{field.label}</Table.CellLabel>
            {getTableFieldValue(field, item)}
          </Table.Cell>
        ))}
        {(updateTrigger || deleteTrigger) && (
          <Table.Cell>
            <Table.CellLabel>{actionsLabel}</Table.CellLabel>
            {updateTrigger && (
              <Button
                modifiers="primary"
                onClick={() => {
                  onUpdateClick(item);
                }}
              >
                {updateTrigger}
              </Button>
            )}{' '}
            {deleteTrigger && (
              <Button
                modifiers="negative"
                onClick={() => {
                  onDeleteClick(item);
                }}
              >
                {deleteTrigger}
              </Button>
            )}
          </Table.Cell>
        )}
      </Table.Row>
    ))}
  </Table.Body>
);

Body.propTypes = {
  actionsLabel: PropTypes.node,
  updateTrigger: PropTypes.node,
  deleteTrigger: PropTypes.node,
  fields: PropTypes.instanceOf(Array),
  items: PropTypes.instanceOf(Array),
  onDeleteClick: PropTypes.func,
  onUpdateClick: PropTypes.func,
};

Body.defaultProps = {
  fields: [],
  items: [],
  actionsLabel: '',
  onDeleteClick: () => {},
  onUpdateClick: () => {},
  updateTrigger: null,
  deleteTrigger: null,
};

export default Body;
