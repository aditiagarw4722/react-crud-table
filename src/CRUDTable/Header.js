// @flow
import * as React from 'react';

import { Table } from './wrappers';
import { chevron, toggleDirection } from './helpers';
import { NO_OP } from '../helpers';
import Select from '../Select';
import { SORT_DIRECTIONS } from './constants';

type Field = {
  name: string,
  label: string,
  sortable?: boolean,
};

type Sort = {
  field: string,
  direction: 'descending' | 'ascending',
};

const mapFieldsToOptions = (fields: Array<Field>) =>
  fields.map((x) => ({
    text: x.label,
    value: x.name,
    key: x.name,
  }));

const sortDirectionsOptions = Object.keys(SORT_DIRECTIONS).map((k) => ({
  text: SORT_DIRECTIONS[k],
  value: SORT_DIRECTIONS[k],
  key: k,
}));

export type Props = {
  onChange: Function,
  actionsLabel: number | string | React.Element<any> | Array<any>,
  fields: Array<Field>,
  sort: Sort,
};

const Header = (props: Props): React.Element<any> => {
  const { fields, sort, onChange, actionsLabel } = props;
  const sortableFields = fields.filter((field) => field.sortable);

  return (
    <Table.Header>
      <Table.Row modifiers="fields">
        {fields.map((field) => (
          <Table.HeaderCell
            key={field.name}
            onClick={() => {
              if (field.sortable) {
                const newDirection = toggleDirection(
                  sort.direction,
                  field.name === sort.field
                );
                onChange(field.name, newDirection);
              }
            }}
          >
            {field.label} {sort.field === field.name && chevron(sort.direction)}
          </Table.HeaderCell>
        ))}
        {actionsLabel && <Table.HeaderCell>{actionsLabel}</Table.HeaderCell>}
      </Table.Row>
      {sortableFields.length > 0 && (
        <Table.Row modifiers="sorter">
          <Table.HeaderCell colSpan={fields.length + +!!actionsLabel}>
            Sort by: <br />
            <Select
              placeholder="Select field"
              options={mapFieldsToOptions(sortableFields)}
              value={sort.field}
              onChange={(evt) => {
                const { value } = evt.currentTarget;
                if (value) {
                  onChange(value, sort.direction);
                }
              }}
            />{' '}
            <Select
              placeholder="Select direction"
              options={sortDirectionsOptions}
              value={sort.direction}
              onChange={(evt) => {
                const { value } = evt.currentTarget;
                if (value) {
                  onChange(sort.field, value);
                }
              }}
            />
          </Table.HeaderCell>
        </Table.Row>
      )}
    </Table.Header>
  );
};

Header.defaultProps = {
  fields: ([]: Array<Object>),
  onChange: NO_OP,
  actionsLabel: '',
};

export default Header;
