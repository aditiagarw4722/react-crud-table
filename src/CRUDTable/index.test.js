import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { create } from 'react-test-renderer';

import CRUDTable, {
  Fields,
  Field,
  CreateForm,
  UpdateForm,
  DeleteForm,
  Pagination,
} from '.';

import Header from './Header';
import Body from './Body';
import PaginationCpt from '../Pagination';
import FormModal from '../FormModal';
import QueryBuilder from '../QueryBuilder';

// eslint-disable-next-line react/prop-types
const DescriptionRenderer = ({ field }) => <textarea {...field} />;

const renderer = new ShallowRenderer();

describe('CRUDTable', () => {
  let tasks;
  let createValidate;
  let updateValidate;
  let deleteValidate;
  let createSubmit;
  let updateSubmit;
  let deleteSubmit;
  let onChange;

  beforeEach(() => {
    tasks = [
      {
        id: 1,
        title: 'Create an example',
        description: 'Create an example of how to use the component',
      },
      {
        id: 2,
        title: 'Improve',
        description: 'Improve the component!',
      },
    ];
    createValidate = jest.fn();
    updateValidate = jest.fn();
    deleteValidate = jest.fn();
    createSubmit = jest.fn(() => Promise.resolve({ success: true }));
    updateSubmit = jest.fn(() => Promise.resolve({ success: true }));
    deleteSubmit = jest.fn(() => Promise.resolve({ success: true }));
    onChange = jest.fn();
  });

  it('renders as expected with forms, query builder, and pagination', () => {
    const result = renderer.render(
      <CRUDTable caption="Tasks" items={tasks} showQueryBuilder>
        <Fields>
          <Field name="id" label="Id" hideInCreateForm readOnly />
          <Field name="title" label="Title" placeholder="Title" />
          <Field
            name="description"
            label="Description"
            render={DescriptionRenderer}
          />
        </Fields>
        <CreateForm
          title="Task Creation"
          message="Create a new task!"
          trigger="Create Task"
          onSubmit={createSubmit}
          submitText="Create"
          validate={createValidate}
        />

        <UpdateForm
          title="Task Update Process"
          message="Update task"
          trigger="Update"
          onSubmit={updateSubmit}
          submitText="Update"
          validate={updateValidate}
        />

        <DeleteForm
          title="Task Delete Process"
          message="Are you sure you want to delete the task?"
          trigger="Delete"
          onSubmit={deleteSubmit}
          submitText="Delete"
          validate={deleteValidate}
        />
        <Pagination itemsPerPage={2} totalOfItems={tasks.length} />
      </CRUDTable>
    );
    expect(result).toMatchSnapshot();
  });

  it('renders as expected without forms, query builder, and pagination', () => {
    const result = renderer.render(
      <CRUDTable caption="Tasks" items={tasks}>
        <Fields>
          <Field name="id" label="Id" hideInCreateForm readOnly />
          <Field name="title" label="Title" placeholder="Title" />
          <Field
            name="description"
            label="Description"
            render={DescriptionRenderer}
          />
        </Fields>
      </CRUDTable>
    );
    expect(result).toMatchSnapshot();
  });

  it('renders as expected when passing promise fetchers', () => {
    const fetchItems = jest.fn(() => Promise.resolve(tasks));
    const fetchTotal = jest.fn(() => Promise.resolve(tasks.length));
    const result = renderer.render(
      <CRUDTable caption="Tasks" fetchItems={fetchItems}>
        <Fields>
          <Field name="id" label="Id" hideInCreateForm readOnly />
          <Field name="title" label="Title" placeholder="Title" />
          <Field
            name="description"
            label="Description"
            render={DescriptionRenderer}
          />
        </Fields>
        <Pagination itemsPerPage={2} fetchTotalOfItems={fetchTotal} />
      </CRUDTable>
    );
    expect(result).toMatchSnapshot();
  });

  describe('Behavior', () => {
    let result;

    beforeEach(() => {
      result = create(
        <CRUDTable
          onChange={onChange}
          caption="Tasks"
          items={tasks}
          showQueryBuilder
        >
          <Fields>
            <Field name="id" label="Id" hideInCreateForm readOnly />
            <Field name="title" label="Title" placeholder="Title" />
            <Field
              name="description"
              label="Description"
              render={DescriptionRenderer}
            />
          </Fields>
          <CreateForm
            title="Task Creation"
            message="Create a new task!"
            trigger="Create Task"
            onSubmit={createSubmit}
            submitText="Create"
            validate={createValidate}
          />

          <UpdateForm
            title="Task Update Process"
            message="Update task"
            trigger="Update"
            onSubmit={updateSubmit}
            submitText="Update"
            validate={updateValidate}
          />

          <DeleteForm
            title="Task Delete Process"
            message="Are you sure you want to delete the task?"
            trigger="Delete"
            onSubmit={deleteSubmit}
            submitText="Delete"
            validate={deleteValidate}
          />
          <Pagination itemsPerPage={2} totalOfItems={tasks.length} />
        </CRUDTable>
      );
    });

    it('loads data using promise fetchers', () => {
      const fetchItems = jest.fn(() => Promise.resolve(tasks));
      const fetchTotal = jest.fn(() => Promise.resolve(tasks.length));
      const result = create(
        <CRUDTable caption="Tasks" fetchItems={fetchItems} onChange={onChange}>
          <Fields>
            <Field name="id" label="Id" hideInCreateForm readOnly />
            <Field name="title" label="Title" placeholder="Title" />
            <Field
              name="description"
              label="Description"
              render={DescriptionRenderer}
            />
          </Fields>
          <Pagination itemsPerPage={2} fetchTotalOfItems={fetchTotal} />
        </CRUDTable>
      );
      expect(fetchItems).toHaveBeenCalledTimes(1);
      expect(fetchTotal).toHaveBeenCalledTimes(1);
    });

    it('handles sort changes', () => {
      const button = result.root.findByType(Header);

      button.props.onClick('name', 'descending');

      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange).toHaveBeenCalledWith({
        pagination: {
          activePage: 1,
          fields: [],
          itemsPerPage: 2,
          totalOfItems: 2,
        },
        queryRules: [],
        sort: { direction: 'descending', field: 'name' },
      });
    });

    it('handles query rules changes', () => {
      const builder = result.root.findByType(QueryBuilder);

      const rule = {
        field: 'name',
        condition: 'CONTAINS',
        value: 'John',
      };

      builder.props.onRuleAdded(rule);
      builder.props.onRuleRemoved(rule);

      const base = {
        pagination: {
          activePage: 1,
          fields: [],
          itemsPerPage: 2,
          totalOfItems: 2,
        },
        sort: { direction: 'descending', field: 'id' },
      };

      expect(onChange).toHaveBeenCalledTimes(2);
      expect(onChange).toHaveBeenCalledWith({
        ...base,
        queryRules: [rule],
      });

      expect(onChange).toHaveBeenCalledWith({
        ...base,
        queryRules: [],
      });
    });

    it('handles page changes', () => {
      const button = result.root.findByType(PaginationCpt);

      button.props.onPageChange(2);

      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange).toHaveBeenCalledWith({
        pagination: {
          activePage: 2,
          fields: [],
          itemsPerPage: 2,
          totalOfItems: 2,
        },
        queryRules: [],
        sort: { direction: 'descending', field: 'id' },
      });
    });

    it('handles item creation', () => {
      const forms = result.root.findAllByType(FormModal);

      const item = { name: 'Name' };
      forms[0].props.onSubmit(item).then(() => {
        expect(createSubmit).toHaveBeenCalledTimes(1);
        expect(createSubmit).toHaveBeenCalledWith(item);
        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenCalledWith({
          pagination: {
            activePage: 1,
            fields: [],
            itemsPerPage: 2,
            totalOfItems: 2,
          },
          queryRules: [],
          sort: { direction: 'descending', field: 'id' },
        });
      });
    });

    it('handles item update', () => {
      const forms = result.root.findAllByType(FormModal);

      const item = { name: 'Name' };
      forms[1].props.onSubmit(item).then(() => {
        expect(updateSubmit).toHaveBeenCalledTimes(1);
        expect(updateSubmit).toHaveBeenCalledWith(item);
        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenCalledWith({
          pagination: {
            activePage: 1,
            fields: [],
            itemsPerPage: 2,
            totalOfItems: 2,
          },
          queryRules: [],
          sort: { direction: 'descending', field: 'id' },
        });
      });
    });

    it('handles item delete', () => {
      const forms = result.root.findAllByType(FormModal);

      const item = { name: 'Name' };
      forms[2].props.onSubmit(item).then(() => {
        expect(deleteSubmit).toHaveBeenCalledTimes(1);
        expect(deleteSubmit).toHaveBeenCalledWith(item);
        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenCalledWith({
          pagination: {
            activePage: 1,
            fields: [],
            itemsPerPage: 2,
            totalOfItems: 2,
          },
          queryRules: [],
          sort: { direction: 'descending', field: 'id' },
        });
      });
    });

    it('handles create form visibility change', () => {
      const forms = result.root.findAllByType(FormModal);
      forms[0].props.onVisibilityChange(true);
      expect(result.root.instance.state.createModalVisible).toEqual(true);
    });

    it('handles update form visibility change', () => {
      const forms = result.root.findAllByType(FormModal);
      forms[1].props.onVisibilityChange(true);
      expect(result.root.instance.state.updateModalVisible).toEqual(true);
    });

    it('handles delete form visibility change', () => {
      const forms = result.root.findAllByType(FormModal);
      forms[2].props.onVisibilityChange(true);
      expect(result.root.instance.state.deleteModalVisible).toEqual(true);
    });

    it('handles update modal trigger process', () => {
      const item = {
        id: 1,
        name: 'Test',
      };
      const body = result.root.findByType(Body);
      body.props.onUpdateClick(item);
      expect(result.root.instance.state.updateModalVisible).toEqual(true);
      expect(result.root.instance.state.updateItem).toEqual(item);
    });

    it('handles delete modal trigger process', () => {
      const item = {
        id: 1,
        name: 'Test',
      };
      const body = result.root.findByType(Body);
      body.props.onDeleteClick(item);
      expect(result.root.instance.state.deleteModalVisible).toEqual(true);
      expect(result.root.instance.state.deleteItem).toEqual(item);
    });
  });
});
