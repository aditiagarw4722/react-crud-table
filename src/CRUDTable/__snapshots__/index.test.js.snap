// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CRUDTable renders as expected when passing promise fetchers 1`] = `
<div>
  <FormModal
    data={
      Object {
        "fields": Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ],
        "message": "Create a new task!",
        "onSubmit": [MockFunction],
        "submitText": "Create",
        "title": "Task Creation",
        "trigger": "Create Task",
        "validate": [MockFunction],
      }
    }
    initialValues={null}
    onSubmit={[Function]}
    onVisibilityChange={[Function]}
    shouldReset={true}
    trigger="Create Task"
    visible={false}
  />
  <Caption>
    Tasks
  </Caption>
  <Table>
    <Header
      actionsLabel="Actions"
      fields={
        Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": true,
            "hideInUpdateForm": false,
            "label": "Id",
            "name": "id",
            "queryable": true,
            "readOnly": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ]
      }
      onChange={[Function]}
      sort={
        Object {
          "direction": "descending",
          "field": "id",
        }
      }
    />
    <Body
      actionsLabel="Actions"
      deleteTrigger="Delete"
      fields={
        Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": true,
            "hideInUpdateForm": false,
            "label": "Id",
            "name": "id",
            "queryable": true,
            "readOnly": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ]
      }
      items={Array []}
      onDeleteClick={[Function]}
      onUpdateClick={[Function]}
      updateTrigger="Update"
    />
  </Table>
  <FormModal
    data={
      Object {
        "fields": Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": true,
            "hideInUpdateForm": false,
            "label": "Id",
            "name": "id",
            "queryable": true,
            "readOnly": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ],
        "message": "Update task",
        "onSubmit": [MockFunction],
        "submitText": "Update",
        "title": "Task Update Process",
        "trigger": "Update",
        "validate": [MockFunction],
      }
    }
    initialValues={Object {}}
    onSubmit={[Function]}
    onVisibilityChange={[Function]}
    shouldReset={false}
    trigger={null}
    visible={false}
  />
  <FormModal
    data={
      Object {
        "fields": Array [],
        "message": "Are you sure you want to delete the task?",
        "onSubmit": [MockFunction],
        "submitText": "Delete",
        "title": "Task Delete Process",
        "trigger": "Delete",
        "validate": [MockFunction],
      }
    }
    initialValues={Object {}}
    onSubmit={[Function]}
    onVisibilityChange={[Function]}
    shouldReset={false}
    trigger={null}
    visible={false}
  />
</div>
`;

exports[`CRUDTable renders as expected with forms, query builder, and pagination 1`] = `
<div>
  <FormModal
    data={
      Object {
        "fields": Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ],
        "message": "Create a new task!",
        "onSubmit": [MockFunction],
        "submitText": "Create",
        "title": "Task Creation",
        "trigger": "Create Task",
        "validate": [MockFunction],
      }
    }
    initialValues={null}
    onSubmit={[Function]}
    onVisibilityChange={[Function]}
    shouldReset={true}
    trigger="Create Task"
    visible={false}
  />
  <Caption>
    Tasks
  </Caption>
  <QueryBuilder
    fields={
      Array [
        Object {
          "hideFromTable": false,
          "hideInCreateForm": true,
          "hideInUpdateForm": false,
          "label": "Id",
          "name": "id",
          "queryable": true,
          "readOnly": true,
          "sortable": true,
          "tableValueResolver": null,
          "type": "text",
          "value": "id",
        },
        Object {
          "hideFromTable": false,
          "hideInCreateForm": false,
          "hideInUpdateForm": false,
          "label": "Title",
          "name": "title",
          "placeholder": "Title",
          "queryable": true,
          "sortable": true,
          "tableValueResolver": null,
          "type": "text",
          "value": "title",
        },
        Object {
          "hideFromTable": false,
          "hideInCreateForm": false,
          "hideInUpdateForm": false,
          "label": "Description",
          "name": "description",
          "queryable": true,
          "render": [Function],
          "sortable": true,
          "tableValueResolver": null,
          "type": "text",
          "value": "description",
        },
      ]
    }
    onRuleAdded={[Function]}
    onRuleRemoved={[Function]}
    queryRules={Array []}
  />
  <Table>
    <Header
      actionsLabel="Actions"
      fields={
        Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": true,
            "hideInUpdateForm": false,
            "label": "Id",
            "name": "id",
            "queryable": true,
            "readOnly": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ]
      }
      onChange={[Function]}
      sort={
        Object {
          "direction": "descending",
          "field": "id",
        }
      }
    />
    <Body
      actionsLabel="Actions"
      deleteTrigger="Delete"
      fields={
        Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": true,
            "hideInUpdateForm": false,
            "label": "Id",
            "name": "id",
            "queryable": true,
            "readOnly": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ]
      }
      items={
        Array [
          Object {
            "description": "Create an example of how to use the component",
            "id": 1,
            "title": "Create an example",
          },
          Object {
            "description": "Improve the component!",
            "id": 2,
            "title": "Improve",
          },
        ]
      }
      onDeleteClick={[Function]}
      onUpdateClick={[Function]}
      updateTrigger="Update"
    />
  </Table>
  <Pagination
    activePage={1}
    fields={Array []}
    itemsPerPage={2}
    onPageChange={[Function]}
    totalOfItems={2}
  />
  <FormModal
    data={
      Object {
        "fields": Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": true,
            "hideInUpdateForm": false,
            "label": "Id",
            "name": "id",
            "queryable": true,
            "readOnly": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ],
        "message": "Update task",
        "onSubmit": [MockFunction],
        "submitText": "Update",
        "title": "Task Update Process",
        "trigger": "Update",
        "validate": [MockFunction],
      }
    }
    initialValues={Object {}}
    onSubmit={[Function]}
    onVisibilityChange={[Function]}
    shouldReset={false}
    trigger={null}
    visible={false}
  />
  <FormModal
    data={
      Object {
        "fields": Array [],
        "message": "Are you sure you want to delete the task?",
        "onSubmit": [MockFunction],
        "submitText": "Delete",
        "title": "Task Delete Process",
        "trigger": "Delete",
        "validate": [MockFunction],
      }
    }
    initialValues={Object {}}
    onSubmit={[Function]}
    onVisibilityChange={[Function]}
    shouldReset={false}
    trigger={null}
    visible={false}
  />
</div>
`;

exports[`CRUDTable renders as expected without forms, query builder, and pagination 1`] = `
<div>
  <FormModal
    data={
      Object {
        "fields": Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ],
        "message": "Create a new task!",
        "onSubmit": [MockFunction],
        "submitText": "Create",
        "title": "Task Creation",
        "trigger": "Create Task",
        "validate": [MockFunction],
      }
    }
    initialValues={null}
    onSubmit={[Function]}
    onVisibilityChange={[Function]}
    shouldReset={true}
    trigger="Create Task"
    visible={false}
  />
  <Caption>
    Tasks
  </Caption>
  <Table>
    <Header
      actionsLabel="Actions"
      fields={
        Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": true,
            "hideInUpdateForm": false,
            "label": "Id",
            "name": "id",
            "queryable": true,
            "readOnly": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ]
      }
      onChange={[Function]}
      sort={
        Object {
          "direction": "descending",
          "field": "id",
        }
      }
    />
    <Body
      actionsLabel="Actions"
      deleteTrigger="Delete"
      fields={
        Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": true,
            "hideInUpdateForm": false,
            "label": "Id",
            "name": "id",
            "queryable": true,
            "readOnly": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ]
      }
      items={
        Array [
          Object {
            "description": "Create an example of how to use the component",
            "id": 1,
            "title": "Create an example",
          },
          Object {
            "description": "Improve the component!",
            "id": 2,
            "title": "Improve",
          },
        ]
      }
      onDeleteClick={[Function]}
      onUpdateClick={[Function]}
      updateTrigger="Update"
    />
  </Table>
  <FormModal
    data={
      Object {
        "fields": Array [
          Object {
            "hideFromTable": false,
            "hideInCreateForm": true,
            "hideInUpdateForm": false,
            "label": "Id",
            "name": "id",
            "queryable": true,
            "readOnly": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Title",
            "name": "title",
            "placeholder": "Title",
            "queryable": true,
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
          Object {
            "hideFromTable": false,
            "hideInCreateForm": false,
            "hideInUpdateForm": false,
            "label": "Description",
            "name": "description",
            "queryable": true,
            "render": [Function],
            "sortable": true,
            "tableValueResolver": null,
            "type": "text",
          },
        ],
        "message": "Update task",
        "onSubmit": [MockFunction],
        "submitText": "Update",
        "title": "Task Update Process",
        "trigger": "Update",
        "validate": [MockFunction],
      }
    }
    initialValues={Object {}}
    onSubmit={[Function]}
    onVisibilityChange={[Function]}
    shouldReset={false}
    trigger={null}
    visible={false}
  />
  <FormModal
    data={
      Object {
        "fields": Array [],
        "message": "Are you sure you want to delete the task?",
        "onSubmit": [MockFunction],
        "submitText": "Delete",
        "title": "Task Delete Process",
        "trigger": "Delete",
        "validate": [MockFunction],
      }
    }
    initialValues={Object {}}
    onSubmit={[Function]}
    onVisibilityChange={[Function]}
    shouldReset={false}
    trigger={null}
    visible={false}
  />
</div>
`;
